effect/ref.py

```diff
@@ -0,0 +1,60 @@
+from characteristic import attributes
+
+from ._base import Effect
+from ._dispatcher import TypeDispatcher
+from ._sync import sync_performer
+
+
+class ERef(object):
+    """
+    An effectful mutable variable.
+
+    Compare to Haskell's ``IORef`` or Clojure's ``atom``.
```

Comment:
Alright, I'm going to try to explain it a bit more in-depth. I'm going to take a big step back and start from the beginning to try to explain how re-assigning a variable and making an HTTP request are the same kind of thing.

Effect has a few different reasons for existing, but each reason can stand on its own.

1. it separates performers from intents, so you can easily and naturally have a library that e.g. supports both blocking and non-blocking backends for an HTTP request.
2. intents are represented transparently so you can easily compare them in your tests
3. everything is purely functional, which has its own benefits for reasoning about code and for testability.

I'm mainly going to talk about the third one, though ERef also benefits from number 2.

It makes just as much sense to have an intent to represent an HTTP request as it does to have an intent to represent a mutable variable. Both of these things are side-effects, and both of them are otherwise trivial to do with typical Python -- you either call `request.get()` or you say `self.x = foo`. They both cause interactions with the outside world -- the request does network I/O, and the mutable variable assignment allows other logical threads of execution to observe some change in its environment.

Effect is a way to represent both of these scenarios with a consistent interface, where instead of making the HTTP request or reassigning the variable in your core logic, you represent those actions with an Effect, and the things that depend on those actions as callbacks to the Effect -- which are all truly pure functions.

So I think if someone is interested in representing their side-effects with Effect, to allow them to write their code purely functional or to make testing their code easy, it makes just as much sense for them to use Effect to represent mutable variables (in the cases where those mutable variables are truly necessary) as for IO.

I hope this helps a bit, and gives more context to this class. I'm happy to improve the ERef docstring more, but I think some of this probably does deserve to be in the more general Effect documentation. Is anything still unclear? If so, I'm very happy to continue discussing. If this explanation was clear enough, then can you offer any feedback on what might improve the ERef docstring?

Thanks for reviewing this and engaging me to come up with clearer descriptions for all this stuff :)